keps/sig-api-machinery/20190621-apply-scalability.md

```diff
@@ -0,0 +1,154 @@
+
+### Store managedFields as a blob
+
+Instead of a highly structured nested map, with lots of small fields, we could store it as a blob. The blob would save on allocation/GC of small objects when deserializing. We could also gzip the entire thing to save space. The downside is that it would make the field harder/impossible for a user to understand. This is mitigated by combining with solution 1, or by client-go/kubectl unzipping the field automatically.
+
+### Convert managedFields directly to internal fieldpath
+
+On updates and applies, the apiserver needs to convert managedFields into an internal representation. We originally decoupled the managedFields API from the internal representation of a fieldset, to allow arbitrary changes to the internal representation without breaking the API. However converting from a serialized config to a gostruct and then into a different struct causes a lot of allocations of small objects and stress on the garbage collector. Instead, we can directly serialize as the internal representation.
+
+The internal format could be stored in protobuf, which would make it much smaller than json (json has a lot of repeated fields that make it extremely large).
+
+### Only update managedFields if previously applied to
+
+Most objects created with post or put are never applied to, so tracking managedFields on them is a waste of resources. If an object without managedFields does get applied to, we can still generate a single managedFields entry for all the existing fields. The generated managedFields conflicts set will be the same, but we won’t have any way of knowing know who is being conflicted with. CPU and memory cost on the apiserver will be dependent on the number of objects managed by server-side apply. It would increase more gradually as controllers decided to adopt SSA. Adoption could be decided on a controller by controller basis, weighing the cost vs. benefits each time. If someone applies to an object created by post or put, the conflicts they see would not contain information about the timestamp or field manager, so conflicts become less useful.
+
+### Storing in one format, serving in another
+
+Another possible solution would be to combine many of these solutions:
+
+Some combination of storing the field managers as one big blob in etcd, using the internal fieldpath representation, as gzipped protobuf, would give us a much lower overhead in etcd and very fast serialization/deserialization. This is especially true if deserialization doesn’t need to know about the specifics of managedfields (hence just deserializing as a blob).
+
```

Comment:
Just as an off the cuff, I think you could get 90% of the benefit by declaring the field a raw extension and adding validation logic to it that does late deserialization.  Use JSON maps (same format) but simply lazily decode it only on clients that care about it.

Reply:
> Just as an off the cuff, I think you could get 90% of the benefit these approaches have by declaring the field a raw extension a

We did this for image metadata in OpenShift.  Client have to know they want it to get it an manipulate it, but it's fairly manageable.  When they mutate, they can place the converted bit into RawExtension.Object and clear RawExtension.Raw and it *should* serialize back properly.

Reply:
>  Client have to know they want it to get it an manipulate it

Does this mean that it is only served when specifically requested? or just that it is more complicated to manipulate because the client has to know how to do the (de)?serialization, so only clients which already know about it can actually change it.

Reply:
No, it's always served.   But deserialization is expensive for two reasons - structure (json boundaries) and memory allocation, and memory allocation dwarfs structure by far (microbenchmark encoding/json.Unmarshal(complex, &json.RawMessage{}) which just verifies it's valid JSON vs decoding to a map or struct and it will be 90% of work).  So we get most of the benefit avoid deserialization except when the client needs it, and otherwise pass it around unmodified.

Since only a limited number of clients actually care, you can define rules that are reasonable for our client types.

Now, other clients will be impacted (hey, why does my python client slow down) and that's definitely a risk, but I think we can justify that if the python client code base gets simpler.  And other languages could if they cared do something simpler.

Reply:
We have basically done this. It seems like increased object size is still a problem.